### ISSUE TITLE
Calls to virtual methods with generic parameters are slower with tiering

### ISSUE THREAD

**Issue by kouvel** [dotnet/coreclr#22555]
See example:
https://github.com/dotnet/coreclr/blob/8ad7d5e57e0eb0308453895693a996f5a93c8e08/tests/src/baseservices/TieredCompilation/TieredVtableMethodTests.cs#L106-L110

The call uses an instantiating stub that does a direct call if the method's code entry point is stable at the time of stub generation, or an indirect call otherwise. With tiering, an indirect call is always used, and this indirect call doesn't seem to be easily prefetched, it's sort of like a dependent load. This could be fixed by making the instantiating stub use a direct call, making the call target backpatchable, and backpatching it when the entry point changes.

**Comment by jkotas** [comment]
Isn't this same underlying problem as https://github.com/dotnet/coreclr/issues/20282 ?

**Comment by kouvel** [comment]
It's similar at a high level. A regular method call would go through a precode of some sort when the code entry point is not stable, and in this case it would do an indirect call through the vtable slot. The latter seems to slower consistently, and since the code is in a stub a fix may be used that is different from a fix for #20282 (though it doesn't have to be different).

**Comment by jkotas** [comment]
It should be easy to change the stub to go through a precode instead of indirection if it helps.

**Comment by kouvel** [comment]
I don't think it would help perf. For a fix for #20282 I was thinking that the backpatching changes could be collected and at some point suspend the runtime and backpatch a bunch of the call targets, something similar could be done for the instantiation stub as well. Alternatively because the instantiation stub has a specific layout the call target can be aligned such that it can be backpatched similarly to dispatch stubs, then it wouldn't need runtime suspension.

**Comment by kouvel** [comment]
It's probably a bit more risky to fix this for 3.0 at this point. Moved to future.
